V9.3 External logins and 2FA for members (#3745)
umbraco/UmbracoDocs: Reference/Security/two-factor-authentication/index.md

---
versionFrom: 9.3.0
keywords: 2fa, security, members
meta.Title: "Two-factor authentication"
meta.Description: "Umbraco users and members supports a two-factor authentication (2FA) abstraction for implementing a 2FA provider of your choice"
---

# Two-factor authentication for Members

> Two-factor authentication (2FA) for Umbraco members are activated by implementing an `ITwoFactorProvider` interface and register the implementation. The implementation can use third-party packages to archive for example support for authentication apps like Microsoft- or Google Authentication App.

Reviewer: ```suggestion
Two-factor authentication (2FA) for Umbraco members is activated by implementing an `ITwoFactorProvider` interface and registering the implementation. The implementation can use third-party packages to archive for example support for authentication apps like Microsoft- or Google Authentication App.
```

> Because Umbraco do not control how the UI is for member login and profile edit, the UI for 2FA is shipped at part of the snippets for macros. These can be used as a starting point, before styling the page like you would like.

Reviewer: ```suggestion
Because Umbraco does not control how the UI is for member login and profile edit, the UI for 2FA is shipped as part of the snippets for macros. These can be used as a starting point, before styling the page as you would like.
```

        {
            var member = _memberService.GetByKey(userOrMemberKey);

            var twoFactorAuthenticator = new TwoFactorAuthenticator();
            SetupCode setupInfo = twoFactorAuthenticator.GenerateSetupCode("My application name", member.Username, secret, false);
            return Task.FromResult<object>(new QrCodeSetupData()
            {
                SetupCode = setupInfo,
                Secret = secret
            });
        }

        /// <summary>
        /// Validated the code and the secret of the user.
        /// </summary>
        public bool ValidateTwoFactorPIN(string secret, string code)
        {
            var twoFactorAuthenticator = new TwoFactorAuthenticator();
            return twoFactorAuthenticator.ValidateTwoFactorPIN(secret, code);
        }

        /// <summary>
        /// Validated the two factor setup
        /// </summary>
        /// <remarks>Called to confirm the setup of two factor on the user. In this case we confirm in the same way as we login by validating the PIN.</remarks>
        public bool ValidateTwoFactorSetup(string secret, string token) => ValidateTwoFactorPIN(secret, token);
    }
}
```

> First we create a model with information required to setup the 2FA provider and then we implement the `ITwoFactorProvider` with use of the `TwoFactorAuthenticator` from the GoogleAuthenticator NuGet package.

Reviewer: ```suggestion
First, we create a model with the information required to set up the 2FA provider and then we implement the `ITwoFactorProvider` with the use of the `TwoFactorAuthenticator` from the GoogleAuthenticator NuGet package.
```

            {
                @using (Html.BeginUmbracoForm<UmbTwoFactorLoginController>(nameof(UmbTwoFactorLoginController.Disable)))
                {
                    <input type="hidden" name="providerName" value="@providerName"/>
                    <button type="submit">Disable @providerName</button>
                }
            }
            else if(setupData is QrCodeSetupData qrCodeSetupData)
            {
                @using (Html.BeginUmbracoForm<UmbTwoFactorLoginController>(nameof(UmbTwoFactorLoginController.ValidateAndSaveSetup)))
                {
                    <h3>Setup @providerName</h3>
                    <img src="@qrCodeSetupData.SetupCode.QrCodeSetupImageUrl"/>
                    <p>Scan the code above with your authenticator app <br /> and enter the resulting code here to validate:</p>
                    <input type="hidden" name="providerName" value="@providerName"  />
                    <input type="hidden" name="secret" value="@qrCodeSetupData.Secret"  />
                    <input type="text" name="code"  />
                    <button type="submit">Validate & save</button>
                }
            }
        }
    }
}
```

> In this razor-code sample, we get the current members unique key and list all registered `ITwoFactorProvider` implementations.
> 
> If the `setupData` is `null` for the specified `providerName` it means the provider is already setup. In this case we show a disable button. Otherwise we check the type and show the UI for how to setup the App Authenticator, by showing the QR Code and a input field to validate the code from the App Authenticator.

Reviewer: ```suggestion
In this razor-code sample, we get the current members unique key and list all registered `ITwoFactorProvider` implementations.

If the `setupData` is `null` for the specified `providerName` it means the provider is already set up. In this case, we show a disable button. Otherwise, we check the type and show the UI for how to set up the App Authenticator, by showing the QR Code and an input field to validate the code from the App Authenticator.
```

> The last part required is to use the `Login` Partial Macro snippet.
> 
> ## Notification when 2FA is requested for a member
> When 2FA login is requested for a member, the `MemberTwoFactorRequestedNotification` is published.
> This notification can also be used to send the member a one time password on e-mail or phone, even that these 2FA types is [not considered as secure](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/mfa?view=aspnetcore-6.0#mfa-sms) as App Authentication, it is still a massive improvement compared to no 2FA.

Reviewer: ```suggestion
The last part required is to use the `Login` Partial Macro snippet.

## Notification when 2FA is requested for a member
When a 2FA login is requested for a member, the `MemberTwoFactorRequestedNotification` is published.
This notification can also be used to send the member a one-time password on e-mail or phone, even though these 2FA types are [not considered as secure](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/mfa?view=aspnetcore-6.0#mfa-sms) as App Authentication, it is still a massive improvement compared to no 2FA.
```
